### DISCUSSION TITLE
Multiple versions of live dependency

### DISCUSSION THREAD

**Discussion by Namedix** [pointfreeco/swift-dependencies#88] [Q&A] [answered]
I would like to achieve various multiple implementations of the same client, let's say ApiClient
```swift
struct ApiClient {
    var request: () -> Data

    func request(...) { ... }
}
```

Currently we can add `live`, `test` and `preview` implementations, but what I would like to achieve is diffent `live` versions of a given client? Ie. for ApiClient, I would like to have one `Authorised` and another `NonAuthorised` version with different implementations.

With environment approach I would just create 2 static live implementations of this `ApiClient` struct like
```swift
extension ApiClient {
    static var authorised: Self { ... }
    static var nonAuthorised: Self { ... }
}
```
but with `swift-dependencies` only option for this is
```swift
extension DependencyValues {
    var apiClient: ApiClient {
        get { self[ApiClient.self] }
        set { self[ApiClient.self] = newValue }
    }
}
```

Is there any way to solve this? or do I need to change implementation of `ApiClient` to fit my needs instead of different implementation of this ~protocol~ ~interface~ struct.
```swift
struct ApiClient {
    var request: () -> Data
    var authorisedRequest: () -> Data

    func request(...) { ... }
    func authorisedRequest(...) { ... }
}
```
---
I know that I can create `authorised` and `nonAuthorised` implementation in extension and then use
```swift
withDependencies {
    $0.apiClient = .authorised
} operation {
    Model()
}
```
 while creating model to change apiClient, but I would like to have it more static not relying on initialisation, but rather on typing `@Dependency(\.authorisedApiClient) var authorisedApiClient` and in other places `@Dependency(\.nonAuthorisedApiClient) var nonAuthorisedApiClient`

**Comment by mbrandonw** [comment, marked as answer]
It sounds like you want _multiple_ versions of `APIClient` available in your application at once? If so, then you are free to add as many computed properties to `DependencyValues` as you want. You just need to create a separate type to act as a key:

```swift
private enum AuthorizedAPIClientKey: DependencyKey { /* ... */ }
extension DependencyValues {
  var authorizedApiClient: ApiClient {
    get { self[AuthorizedAPIClientKey.self] }
    set { self[AuthorizedAPIClientKey.self] = newValue }
  }
}

private enum NonAuthorizedAPIClientKey: DependencyKey { /* ... */ }
extension DependencyValues {
  var nonAuthorizedApiClient: ApiClient {
    get { self[NonAuthorizedAPIClientKey.self] }
    set { self[NonAuthorizedAPIClientKey.self] = newValue }
  }
}
```

**Reply by theMuslimDev** [reply]
How can I have the multiple "versions" of the a dependency that I own? Let's say I have a class called `SomeClass`. In order to register this dependency I can just have an extension to my class:

```swift
extension SomeClass: DependencyKey {
  static let liveValue: FirstProtocol = SomeClass()
}

extension DependencyValues {
  var apiClient: FirstProtocol {
    get { self[SomeClass.self] }
    set { self[SomeClass.self] = newValue }
  }
}
```

How will I create another live version of this dependency that conforms to another protocol?

**Reply by Namedix** [reply]
@mbrandonw Oh! I didn't know that we can use different type for `liveValue` in `DependencyKey` Thats changes a lot!

For others with my example we can do it like this:

Type
```swift
struct ApiClient {
    var request: () -> Data

    func request(...) { ... }
}
```
Non-auth ApiClient 
```swift
extension ApiClient: DependencyKey {
    static var liveValue: ApiClient { /* Non authorized request implementation */ }
}

extension DependencyValues {
    var apiClient: ApiClient {
        get { self[ApiClient.self] }
        set { self[ApiClient.self] = newValue }
    }
}
```
Authorized ApiClient
```swift
private enum AuthorizedAPIClientKey: DependencyKey {
    static var liveValue: ApiClient { /* Authorized request implementation */ }
}

extension DependencyValues {
  var authorizedApiClient: ApiClient {
    get { self[AuthorizedAPIClientKey.self] }
    set { self[AuthorizedAPIClientKey.self] = newValue }
  }
}
```

**Reply by mbrandonw** [reply]
@Namedix Yeah, it's a shortcut to use the dependency type as the thing that conforms to `DependencyKey`, but _any_ type can conform. This is how SwiftUI environment values typically work. You usually define a little private, empty enum and conform it to `EnvironmentKey`.

@Muhammed9991 I believe my answer above answers your question. Rather than conforming `SomeClass` to `DependencyKey` you can create some private, empty enum types that represent the various versions of the dependency and conform them to `DependencyKey`.

**Reply by theMuslimDev** [reply]
@mbrandonw Sorry I don't quite understand how that would work. Is it possible to give a simple example for that?

**Reply by mbrandonw** [reply]
@Muhammed9991 Perhaps I don't understand your example. Can you expand a bit more?

**Reply by theMuslimDev** [reply]
For context, I have some dependencies currently written using Factory. I'm trying to move them to use `@Dependency` instead.

In Factory, I'll do something like this:

```swift
var apiClientOne: Factory<FirstProtocol> { self { SomeClass() } }
var apiClientTwo: Factory<SecondProtocol> { self { SomeClass() } }
```

I'm not sure what the equivalent in `swift-dependencies` will be. Because the dependencies are essentially the same underlying class but different protocol conformance. So I can't do something like this:

```swift
extension SomeClass: DependencyKey {
    static let liveValue1: FirstProtocol = SomeClass()
    static let liveValue2: SecondProtocol = SomeClass()
}

extension DependencyValues {
    var apiClientOne: FirstProtocol {
        get { self[SomeClass.self] }
        set { self[SomeClass.self] = newValue }
    }
}

extension DependencyValues {
    var apiClientTwo: SecondProtocol {
        get { self[SomeClass.self] }
        set { self[SomeClass.self] = newValue }
    }
}
```

Just a side note. This protocol oriented approach reminds of `Protocol witness` series by PointFree 😄.

**Reply by stephencelis** [reply]
@Muhammed9991 Conforming the dependency itself to `DependencyKey` is a convenience but not necessary. You can define unique types to the protocol and still delegate to the same class as a dependency:

```swift
enum APIClientOneKey: DependencyKey {
  static let liveValue: FirstProtocol = SomeClass()
}
enum APIClientTwoKey: DependencyKey {
  static let liveValue: SecondProtocol = SomeClass()
}

extension DependencyValues {
  var apiClientOne: FirstProtocol {
    get { self[APIClientOneKey.self] }
    set { self[APIClientOneKey.self] = newValue }
  }
}

extension DependencyValues {
  var apiClientTwo: SecondProtocol {
    get { self[APIClientTwoKey.self] }
    set { self[APIClientTwoKey.self] = newValue }
  }
}
```

**Reply by mbrandonw** [reply]
Hi @Muhammed9991, I believe your situation does fit into what I was trying to explain above. You have a single object that conforms to two protocols, and you want to expose that as two separate dependencies, one for each protocol:

```swift
class SomeClass: FirstProtocol, SecondProtocol { /* ... */ }

private enum FirstDependencyKey: DependencyKey {
  static let liveValue: any FirstProtocol = SomeClass()
}
extension DependencyValues {
 var apiClientOne: FirstProtocol {
    get { self[FirstDependencyKey.self] }
    set { self[FirstDependencyKey.self] = newValue }
  }
}

private enum SecondDependencyKey: DependencyKey {
  static let liveValue: any SecondProtocol = SomeClass()
}
extension DependencyValues {
 var apiClientTwo: SecondProtocol {
    get { self[SecondDependencyKey.self] }
    set { self[SecondDependencyKey.self] = newValue }
  }
}
```

Note that it is not necessary to conform your dependency object to `DependencyKey` directly. That's just a shortcut that can be handy for most situations. Any type can conform to the `DependencyKey` protocol, and often it can just be a little private, empty enum to expose the live value. This is exactly how SwiftUI's environment works.

**Reply by theMuslimDev** [reply]
Ahh that makes sense. I think I was over complicating it.
